### PR TITLE
feat: block player movement

### PR DESCRIPTION
**Pull request by AlejandroAlvarezMelucciDCL**
This PR is part of [Cycle's 8 Shape Feature block player movement](https://www.notion.so/decentraland/Feature-block-player-movement-f31a686e41a84246b0e996a44e648731)

Related PRs:
SDK: [#991](https://github.com/decentraland/js-sdk-toolchain/pull/991)
Protocol: [#215](https://github.com/decentraland/protocol/pull/215)

## QA TEST INSTRUCTIONS

* Download this PR's Explorer build
* Once in Goerli Plaza go to the deployed scene with `/goto -132,-42`
* Once the scene loads you are ready to test 
    * On the right there's a scene UI with buttons, these are toggles, red for disabled, green for enabled.
    * The options can be used in different combinations (2 or more!), like Disable Walk and Disable Jump.
    * Disable All takes precedence, so if that's enabled, it doesn't matter the rest, but disabling it keeps the state of the other toggles.
    * Test going in and out the scene, walking, running, jogging.
    * Test standing on the moving platform and disable different motions.
    * The cube in the middle triggers an emote when interacted with (mouse click), this is an emote triggered by the scene, disabling emote in the scene UI won't prevent the cube from triggering the emote but will prevent the user to interrupt it. Think of this like "pulling a level" animation that the scene creator don't want the player to interrupt.
image

## Summary by CodeRabbit

- **New Features**
	- Introduced input modification capabilities through the new `InputModifierComponent`, allowing for dynamic control over player actions such as walking, running, and jumping.
	- Added the `InputModifierPlugin` to enhance input management within the game environment.
	- New `InputModifierHandlerSystem` to manage input modifiers based on scene changes.
	- Expanded component identifiers with new constants for camera management and input handling.
	- Introduced new protocol buffer classes for managing camera transitions and main camera functionality.
	- Enhanced test setup to include `InputModifierComponent`, allowing for nuanced input behavior in jump mechanics.
	- Updated dependencies to include the latest version of `@dcl/protocol` and added `@dcl/sdk`.

- **Bug Fixes**
	- Improved handling of unintended emote activations by integrating input modifier checks.

- **Documentation**
	- Updated metadata files for new components and systems to ensure proper asset management.

### CODE REVIEW

File: scripts/package-lock.json
Comment (AlejandroAlvarezMelucciDCL) [inline scripts/package-lock.json:1]: This file will be regenerated before merging, once the PR in `protocol` and `js-sdk-toolchain` are merged.

**Comment by AlejandroAlvarezMelucciDCL** [comment]
I'm done for the day, tomorrow I'll work on the conflicts with main

File: Explorer/Assets/DCL/Input/Systems/InputBlock.cs [hunk 1]
```diff
@@ -20,15 +19,17 @@ public InputBlock(ObjectProxy<DCLInput> dclInput, ObjectProxy<World> globalWorld
 
         public void BlockMovement()
         {
-            globalWorld.StrictObject.AddOrGet(playerEntity.StrictObject, new MovementBlockerComponent());
+            ref var inputModifier = ref globalWorld.StrictObject.Get<InputModifierComponent>(playerEntity.StrictObject);
+            inputModifier.DisableAll = true;
```
[the remark this answers is not in the document]
Comment (AlejandroAlvarezMelucciDCL) [reply Explorer/Assets/DCL/Input/Systems/InputBlock.cs:23]: `InputModifier` is a struct and you can't have a ref field in C# 8

File: Explorer/Assets/DCL/PluginSystem/World/InputModifierPlugin.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        public void InjectToWorld(ref ArchSystemsWorldBuilder<Arch.Core.World> builder, in ECSWorldInstanceSharedDependencies sharedDependencies,
+            in PersistentEntities persistentEntities, List<IFinalizeWorldSystem> finalizeWorldSystems, List<ISceneIsCurrentListener> sceneIsCurrentListeners)
+        {
+            ResetDirtyFlagSystem<PBInputModifier>.InjectToWorld(ref builder);
+            var system = InputModifierHandlerSystem.InjectToWorld(ref builder, globalWorldProxy, playerEntity, sharedDependencies.SceneStateProvider);
+            sceneIsCurrentListeners.Add(system);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (AlejandroAlvarezMelucciDCL) [reply Explorer/Assets/DCL/PluginSystem/World/InputModifierPlugin.cs:41]: The lists are properly reused by the caller of `InjectToWorld` and this method is called only once in the app life cycle

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdateInputMovementSystem.cs [hunk 3]
```diff
@@ -29,38 +30,64 @@ public UpdateInputMovementSystem(World world, DCLInput dclInput) : base(world)
...
+            if (inputModifierComponent.DisableJog)
+            {
+                if (inputModifierComponent.DisableWalk)
+                    return MovementKind.Run;
+
+                return MovementKind.Walk;
+            }
+
+            return MovementKind.Jog;
+        }
```
[the remark this answers is not in the document]
Comment (AlejandroAlvarezMelucciDCL) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdateInputMovementSystem.cs:91]: I think my version is easier to follow, it's a rather entangled logic with all the combinations and I prefer the if statements to be as explicit as possible.

**Review by mikhail-dcl — APPROVED** [review]
I was closely following the PR and can approve it with piece of mind.

GJ 💚

**Pull request merged**